TF Lite GPU delegate support for desktop Linux systems

sfalexrog opened the pull request:
Rationale: Since OpenGL ES 3.1 is widely available for desktop and "sort-of-desktop" computers (like Raspberry Pi 4, Asus Tinkerboard), I believe it would be beneficial for TF Lite GPU delegate to be available there. Basically, I'm moved by the same thoughts as presented in P/R #36180.

I believe most of the code should be usable on "desktop" systems as-is, the only problems were due to `<EGL.h>` pulling in `<X.h>` and `<Xlib.h>` with their defines. I also changed the `egl_sync.cc` code to load EGL functions dynamically instead of using `EGL_EGLEXT_PROTOTYPES` - the latter was causing the library to not load due to missing symbols in Mesa's/GLVND's EGL.

I've tried to stick to the Bazel build system, and even that is basically used to add `libEGL.so` and `libGLESv2.so` to the list of linked libraries for the delegate. The option to link these libraries is `--config=gles`.

sfalexrog committed:
tflite/gpu: Undef some symbols defined by X.h/Xlib.h

These definitions break the build in interesting ways.

sfalexrog committed:
tflite/gpu: Use portable_egl/portable_gl31 wherever possible

sfalexrog committed:
tflite/gpu: Dynamically load EGL_KHR_fence_sync in egl_sync.cc

sfalexrog committed:
tflite/gpu: Don't define EGL_EGLEXT_PROTOTYPES

sfalexrog committed:
tflite/gpu: Add config option for EGL/GLES library linkage

jdduke commented:
@impjdi can you advise on the EGL-specific changes?

impjdi commented on .bazelrc:
```diff
@@ -53,6 +53,7 @@
 #     nogcp:        Disable GCS support.
 #     nohdfs:       Disable hadoop hdfs support.
 #     nonccl:       Disable nccl support.
+#     gles:         Build with OpenGL ES support (TF Lite only).
```
I'm not fan of adding a flag at the global level.  However, read on to my next comment.

sfalexrog replied:
Basically, the only thing that this flag does is add `-lEGL -lGLESv2` to shared libraries. I am okay with dropping the config target, provided I can add these flags for linux targets. I am not familiar with bazel, though, maybe there's a way I can add them without adding a config flag?

impjdi replied:
Sorry, missed this comment of yours.

Yes, you can easily add --linkopt -lEGL --linkopt -lGLESv2 when running bazel in command line, or you can add linkopts = ["-lEGL", "-lGLESv2"] in your BUILD file.  If you grep for lEGL or lGLES in BUILD files, you might be able to find examples.  You can just add those to your terminal cc_library or cc_binary that employs the GPU delegate on your end.

sfalexrog replied:
So if I understand correctly, the only way right now for non-mobile targets to use the GPU delegate is to have a binary built alongside the tensorflow source code? I guess this makes sense, with a lack of an easy way to get a "packaged" library with headers.

Okay, guess this makes sense, though loading the delegate from Python will probably be somewhat complicated (then again, the API is marked experimental for a reason, and the client code may just preload the required libraries).

impjdi commented on tensorflow/lite/delegates/gpu/gl/portable_egl.h:
```diff
@@ -19,4 +19,12 @@ limitations under the License.
 #include <EGL/egl.h>
 #include <EGL/eglext.h>
 
+// Workaround for the awful decisions made in X11
```
This is a wrong workaround.  What you want is

`--copt -DMESA_EGL_NO_X11_HEADERS --copt -DEGL_NO_X11`

and if this is the only thing that you really need, I wonder whether a global define is really necessary...

sfalexrog replied:
That is exactly what I've thought, but my current `EGL/eglplatform.h` does not honor `EGL_NO_X11` (my `eglplatform.h` comes from `libegl1-mesa-dev`, version `1:20.0~git200108124800.ad4fb7e~b~padoka0` - I assume older mesa will not honor `EGL_NO_X11`)

sfalexrog replied:
For the record, the preprocessor check that guards the `<X11/Xlib.h>` and `<X11/Xutil.h>` includes on my system (which is Ubuntu 18.04 + Padoka PPA) is
```C
#elif defined(__unix__) || defined(USE_X11)
```
(so even having `__unix__` defined would lead to X11 headers being included)

sfalexrog replied:
Sorry, turns out the problem is on my side due to using unofficial mesa build (that did not have either `EGL_NO_X11` or `MESA_EGL_NO_X11_HEADERS` checks in place). I've reverted the `portable_egl.h` and `portable_gl31.h` to their previous versions.

sfalexrog committed:
tflite/gpu: Use EGL_NO_X11/MESA_EGL_NO_X11_HEADERS

The workaround that was in place before was mostly due to me having a weird set of EGL headers. Newer EGL.h should have EGL_NO_X11 ifdef, older ones may have MESA_EGL_NO_X11_HEADERS

impjdi commented on tensorflow/lite/delegates/gpu/cl/egl_sync.cc:
```diff
@@ -21,9 +21,49 @@ namespace tflite {
 
+namespace {
+
+PFNEGLCREATESYNCKHRPROC g_eglCreateSyncKHR = nullptr;
```
i have a couple issues with these:
1. global variables, and 
2. no mutex

How often is IsEglSyncSupported called and how slow is eglGetProcAddress?  If it's rare and not performance critical, I would just make it simple without global variables.

sfalexrog replied:
I've tried to basically replicate the code from gl_interop.cc: https://github.com/tensorflow/tensorflow/blob/b780f710ce79cdabef13da19608c80f83f45141c/tensorflow/lite/delegates/gpu/cl/gl_interop.cc#L63-L74
As far as I understand, static initializers are guaranteed to be thread-safe since C++11, though I may be misunderstanding how that deals with the side effects.
`eglGetProcAddress` may have a non-trivial run time, depending on the vendor. In any case, loading functions should probably be performed at initialization time, so maybe I should move this to `opencl_wrapper.cc/.h`?

impjdi replied:
Ah missed the static there.  ok.

impjdi commented on tensorflow/lite/delegates/gpu/cl/egl_sync.cc:
```diff
@@ -21,9 +21,49 @@ namespace tflite {
+  return isAvailable;
+}
+
+#define ENSURE_SYNC_AVALIABLE_ON(display)                           \
```
rather than introducing macro which is not encouraged by the style guide, I would just make the above function return absl::Status instead of bool and get rid of this macro.  you can use RETURN_IF_ERROR(YourFunction(display));

sfalexrog replied:
Thanks, did that!

sfalexrog committed:
tflite/gpu: Use RETURN_IF_ERROR when checking for EGL_KHR_fence_sync

impjdi commented on tensorflow/lite/delegates/gpu/cl/egl_sync.cc:
```diff
@@ -21,9 +21,45 @@ namespace tflite {
+PFNEGLCLIENTWAITSYNCKHRPROC g_eglClientWaitSyncKHR = nullptr;
+
+absl::Status IsEglSyncSupported(EGLDisplay display) {
+  static bool isAvailable = [display]() -> bool {
```
change `isAvailable` to `is_available` or `supported` (not camel case)

sfalexrog replied:
Done!

sfalexrog committed:
tflite/gpu: Drop camel case for internal variable

sfalexrog committed:
tflite/gpu: Revert adding config option for EGL/GLES library linkage

Building is still possible with `--copt -DMESA_EGL_NO_X11_HEADERS --copt -DEGL_NO_X11`, but the resulting binary won't be linked against
EGL/GLES libraries. Thus, loading these libraries before attempting to load EGL delegate or linking against those libraries is the responsibility
of the end user.

impjdi commented on tensorflow/BUILD:
```diff
@@ -272,6 +272,12 @@ config_setting(
     visibility = ["//visibility:public"],
 )
 
+config_setting(
```
i think this is a leftover?

sfalexrog replied:
Yes, didn't notice that was still there. Thanks!

impjdi commented on tensorflow/lite/delegates/gpu/cl/BUILD:
```diff
@@ -208,9 +208,6 @@ cc_library(
     name = "egl_sync",
     srcs = ["egl_sync.cc"],
     hdrs = ["egl_sync.h"],
-    defines = [
```
can you elaborate why this was removed?

sfalexrog replied:
This was causing linker errors (and that's basically the reason I've switched to loading the `EGL_KHR_fence_sync` functions dynamically) since not all implementations expose the "*KHR" functions. Now that the extension functions are dynamically loaded, is there really a reason to keep this define around?

sfalexrog committed:
tensorflow/BUILD: remove gles leftovers

gbaned commented:
@sfalexrog  Can you please resolve conflicts? Thanks!

sfalexrog commented:
Looks like most of what this PR has boiled down to is already implemented in the master branch (specifically, by this commit: https://github.com/tensorflow/tensorflow/commit/7280cb31337d73a2fc869d7ba7cdcaf5295ad26e), so I believe it is okay to close this without merging.

BenjaminWegener commented:
https://github.com/tensorflow/tensorflow/issues/55522
any hints on running tflite gpu/gl on desktop via python? arm64 or x64?

BenjaminWegener commented:
managed to build it after reading this and other issues,
https://github.com/BenjaminWegener/tensorflow_lite_gpu_python/commit/b1fcd1b51c039fca0408eccf49cea77d3bbee2c5
https://github.com/BenjaminWegener/tensorflow_lite_gpu_python/commit/9721986c8289c3fce9642526ef7bb799ade5466e
